exercises/dnd-character/canonical-data.json

```diff
@@ -0,0 +1,176 @@
+        {
+          "description": "ability modifier for score 5 is -3",
+          "property": "modifier",
+          "input": {
+            "score": 5
+          },
+          "expected": -3
+        },
+        {
+          "description": "ability modifier for score 6 is -2",
+          "property": "modifier",
+          "input": {
+            "score": 6
+          },
+          "expected": -2
+        },
+        {
+          "description": "ability modifier for score 7 is -2",
+          "property": "modifier",
+          "input": {
+            "score": 7
+          },
+          "expected": -2
+        },
+        {
+          "description": "ability modifier for score 8 is -1",
+          "property": "modifier",
+          "input": {
+            "score": 8
+          },
+          "expected": -1
+        },
+        {
+          "description": "ability modifier for score 9 is -1",
+          "property": "modifier",
+          "input": {
+            "score": 9
+          },
+          "expected": -1
+        },
+        {
+          "description": "ability modifier for score 10 is 0",
+          "property": "modifier",
+          "input": {
+            "score": 10
+          },
+          "expected": 0
+        },
+        {
+          "description": "ability modifier for score 11 is 0",
+          "property": "modifier",
+          "input": {
+            "score": 11
+          },
+          "expected": 0
+        },
+        {
+          "description": "ability modifier for score 12 is +1",
+          "property": "modifier",
+          "input": {
+            "score": 12
+          },
+          "expected": 1
+        },
+        {
+          "description": "ability modifier for score 13 is +1",
+          "property": "modifier",
+          "input": {
+            "score": 13
+          },
+          "expected": 1
+        },
+        {
+          "description": "ability modifier for score 14 is +2",
+          "property": "modifier",
+          "input": {
+            "score": 14
+          },
+          "expected": 2
+        },
+        {
+          "description": "ability modifier for score 15 is +2",
+          "property": "modifier",
+          "input": {
+            "score": 15
+          },
+          "expected": 2
+        },
+        {
+          "description": "ability modifier for score 16 is +3",
+          "property": "modifier",
+          "input": {
+            "score": 16
+          },
+          "expected": 3
+        },
+        {
+          "description": "ability modifier for score 17 is +3",
+          "property": "modifier",
+          "input": {
+            "score": 17
+          },
+          "expected": 3
+        },
+        {
+          "description": "ability modifier for score 18 is +4",
+          "property": "modifier",
+          "input": {
+            "score": 18
+          },
+          "expected": 4
+        }
+      ]
+    },
+    {
+      "description": "random ability is within range",
```

Comment:
But do language tracks actually implement [randomness tests](https://en.wikipedia.org/wiki/Randomness_tests)?

I can see that the [C# track](https://github.com/exercism/csharp/blob/master/exercises/diffie-hellman/DiffieHellmanTest.cs#L21) checks if the output is distinct between runs. If we don't expect to test the ephemeral quality of randomness, how about we just let it be up to the mentor to validate if the student initialized the random generator in the right way?

Reply:
Since `ability` can only output values in the range 3-18, duplicates are quite likely.

The C# Diffie-Hellman "unique implies random" property relies on the odds 1 : 10^*d* for *d* digits in the private key; I don't know how much *d* is in this exercise, but the point with the exercise is that it's pretty big in practice.

I don't want to make it too hard for track implementers to write tests.

I think what might happen in a lot of cases is that the RNG can be mocked.

Reply:
Real statistical randomness tests are non-trivial to implement. I can't speak for other tracks, but in the Rust track, the [`diffie-hellman` tests](https://github.com/exercism/rust/blob/master/exercises/diffie-hellman/tests/diffie-hellman.rs) just kind of punted on randomness checking. 

Range tests are more plausible to request of non-mathematician test implementers.

Reply:
I think we should defer this addition to a separate issue of testing whether randomness is used as expected, since it may cover more than one exercise (and in particular also the diffie-hellman exercise).

Reply:
Agreed, those are very valid points. Let's defer the discussion then.